Задание 1
Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics


In [90]:
import numpy as np
import pandas as pd

In [91]:
from sklearn.datasets import load_boston

In [92]:
boston = load_boston()

In [93]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [94]:
feature_names = boston.feature_names

In [95]:
X = pd.DataFrame(boston.data, columns = feature_names)

In [96]:
target = boston.target

In [97]:
y = pd.DataFrame(target, columns=['price'])

In [98]:
from sklearn.model_selection import train_test_split

In [99]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [100]:
from sklearn.linear_model import LinearRegression

In [101]:
lr = LinearRegression()

In [102]:
lr.fit(X_train, y_train)

LinearRegression()

In [103]:
y_pred = lr.predict(X_test)

In [104]:
from sklearn.metrics import r2_score

In [105]:
r2_score(y_test, y_pred)

0.7112260057484929

Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.
Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [106]:
from sklearn.ensemble import RandomForestRegressor

In [107]:
model = RandomForestRegressor(n_estimators = 1000, max_depth = 12, random_state = 42)

In [108]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [109]:
y_pred = model.predict(X_test)

In [110]:
r2_score(y_test, y_pred)

0.87472606157312

##### r2_score при линейной регрессии составило 0.7112260057484929 при RandomForestRegressor 0.87472606157312, следовательно RandomForestRegressor работает лучше

*Задание 3
Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.
С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.


In [111]:
?RandomForestRegressor

In [72]:
model.feature_importances_

array([0.03167574, 0.00154252, 0.00713813, 0.00123624, 0.01426897,
       0.40268179, 0.01429864, 0.06397257, 0.00528122, 0.01152493,
       0.01808108, 0.01245085, 0.41584732])

In [73]:
X_pd = pd

array([[2.98500e-02, 0.00000e+00, 2.18000e+00, ..., 1.87000e+01,
        3.94120e+02, 5.21000e+00],
       [1.31580e-01, 0.00000e+00, 1.00100e+01, ..., 1.78000e+01,
        3.93300e+02, 1.20400e+01],
       [1.71420e-01, 0.00000e+00, 6.91000e+00, ..., 1.79000e+01,
        3.96900e+02, 1.02100e+01],
       ...,
       [1.50100e-02, 8.00000e+01, 2.01000e+00, ..., 1.70000e+01,
        3.90940e+02, 5.99000e+00],
       [1.11604e+01, 0.00000e+00, 1.81000e+01, ..., 2.02000e+01,
        1.09850e+02, 2.32700e+01],
       [2.28760e-01, 0.00000e+00, 8.56000e+00, ..., 2.09000e+01,
        7.08000e+01, 1.06300e+01]])

In [118]:
pd.DataFrame({'value': model.feature_importances_}, index = X_train.columns).sort_values(by='value')

value
CHAS     0.001236
ZN       0.001543
RAD      0.005281
INDUS    0.007138
TAX      0.011525
B        0.012451
NOX      0.014269
AGE      0.014299
PTRATIO  0.018081
CRIM     0.031676
DIS      0.063973
RM       0.402682
LSTAT    0.415847

Отсюда видим что два наиболее влияющих показателя на стоимость квартиры: RM и LSTAT

*Задание 4
В этом задании мы будем работать с датасетом, с которым мы уже знакомы по домашнему заданию по библиотеке Matplotlib, это датасет Credit Card Fraud Detection.Для этого датасета мы будем решать задачу классификации - будем определять,какие из транзакциции по кредитной карте являются мошенническими.Данный датасет сильно несбалансирован (так как случаи мошенничества относительно редки),так что применение метрики accuracy не принесет пользы и не поможет выбрать лучшую модель.Мы будем вычислять AUC, то есть площадь под кривой ROC.
Импортируйте из соответствующих модулей RandomForestClassifier, GridSearchCV и train_test_split.
Загрузите датасет creditcard.csv и создайте датафрейм df.
С помощью метода value_counts с аргументом normalize=True убедитесь в том, что выборка несбалансирована. Используя метод info, проверьте, все ли столбцы содержат числовые данные и нет ли в них пропусков.Примените следующую настройку, чтобы можно было просматривать все столбцы датафрейма:
pd.options.display.max_columns = 100.
Просмотрите первые 10 строк датафрейма df.
Создайте датафрейм X из датафрейма df, исключив столбец Class.
Создайте объект Series под названием y из столбца Class.
Разбейте X и y на тренировочный и тестовый наборы данных при помощи функции train_test_split, используя аргументы: test_size=0.3, random_state=100, stratify=y.
У вас должны получиться объекты X_train, X_test, y_train и y_test.
Просмотрите информацию о их форме.
Для поиска по сетке параметров задайте такие параметры:
parameters = [{'n_estimators': [10, 15],
'max_features': np.arange(3, 5),
'max_depth': np.arange(4, 7)}]
Создайте модель GridSearchCV со следующими аргументами:
estimator=RandomForestClassifier(random_state=100),
param_grid=parameters,
scoring='roc_auc',
cv=3.
Обучите модель на тренировочном наборе данных (может занять несколько минут).
Просмотрите параметры лучшей модели с помощью атрибута best_params_.
Предскажите вероятности классов с помощью полученнной модели и метода predict_proba.
Из полученного результата (массив Numpy) выберите столбец с индексом 1 (вероятность класса 1) и запишите в массив y_pred_proba. Из модуля sklearn.metrics импортируйте метрику roc_auc_score.
Вычислите AUC на тестовых данных и сравните с результатом,полученным на тренировочных данных, используя в качестве аргументов массивы y_test и y_pred_proba.


In [119]:
from sklearn.ensemble import RandomForestClassifier

In [120]:
from sklearn.model_selection import GridSearchCV

In [121]:
df = pd.read_csv('creditcard.csv')

In [122]:
def reduce_mem_usage(df):
    start_mem = df.memory_usage().sum() / 1024**2
    print('Memory usage of dataframe is {:.2f} MB'.format(start_mem))

    for col in df.columns:
        col_type = df[col].dtype
        if col_type != object:
            c_min = df[col].min()
            c_max = df[col].max()
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.uint8).min and c_max < np.iinfo(np.uint8).max:
                    df[col] = df[col].astype(np.uint8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.uint16).min and c_max < np.iinfo(np.uint16).max:
                    df[col] = df[col].astype(np.uint16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.uint32).min and c_max < np.iinfo(np.uint32).max:
                    df[col] = df[col].astype(np.uint32)                    
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)
                elif c_min > np.iinfo(np.uint64).min and c_max < np.iinfo(np.uint64).max:
                    df[col] = df[col].astype(np.uint64)
            else:
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)

    end_mem = df.memory_usage().sum() / 1024**2
    print('Memory usage after optimization is: {:.2f} MB'.format(end_mem))
    print('Decreased by {:.1f}%'.format(100 * (start_mem - end_mem) / start_mem))
    return df

In [123]:
reduce_mem_usage(df)

Memory usage of dataframe is 67.36 MB
Memory usage after optimization is: 17.11 MB
Decreased by 74.6%


Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359375  -0.072754  2.537109  1.377930 -0.338379   
1            0.0   1.191406   0.266113  0.166504  0.448242  0.060028   
2            1.0  -1.358398  -1.339844  1.773438  0.379883 -0.503418   
3            1.0  -0.966309  -0.185181  1.792969 -0.863281 -0.010307   
4            2.0  -1.158203   0.877930  1.548828  0.403076 -0.407227   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.882812  10.070312 -9.835938 -2.066406 -5.363281   
284803  172787.0  -0.732910  -0.055084  2.035156 -0.738770  0.868164   
284804  172788.0   1.919922  -0.301270 -3.250000 -0.557617  2.630859   
284805  172788.0  -0.240479   0.530273  0.702637  0.689941 -0.377930   
284806  172792.0  -0.533203  -0.189697  0.703125 -0.506348 -0.012543   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462402  0.239624  0.098694  0.363770  ... -0.018311  0.277832   
1      -0.082336 -0.078796  0.085083 -0.255371  ... -0.225830 -0.638672   
2       1.800781  0.791504  0.247681 -1.514648  ...  0.248047  0.771484   
3       1.247070  0.237549  0.377441 -1.386719  ... -0.108276  0.005272   
4       0.095947  0.592773 -0.270508  0.817871  ... -0.009430  0.798340   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.607422 -4.917969  7.304688  1.914062  ...  0.213501  0.111877   
284803  1.058594  0.024323  0.294922  0.584961  ...  0.214233  0.924316   
284804  3.031250 -0.296875  0.708496  0.432373  ...  0.232056  0.578125   
284805  0.623535 -0.686035  0.679199  0.392090  ...  0.265137  0.799805   
284806 -0.649414  1.577148 -0.414551  0.486084  ...  0.260986  0.643066   

             V23       V24       V25       V26       V27       V28  \
0      -0.110474  0.066956  0.128540 -0.189087  0.133545 -0.021057   
1       0.101318 -0.339844  0.167114  0.125854 -0.008980  0.014725   
2       0.909180 -0.689453 -0.327637 -0.139038 -0.055359 -0.059753   
3      -0.190308 -1.175781  0.647461 -0.221924  0.062744  0.061462   
4      -0.137451  0.141235 -0.206055  0.502441  0.219482  0.215210   
...          ...       ...       ...       ...       ...       ...   
284802  1.014648 -0.509277  1.436523  0.250000  0.943848  0.823730   
284803  0.012466 -1.016602 -0.606445 -0.395264  0.068481 -0.053528   
284804 -0.037506  0.640137  0.265625 -0.087341  0.004456 -0.026566   
284805 -0.163330  0.123230 -0.569336  0.546875  0.108826  0.104553   
284806  0.376709  0.008797 -0.473633 -0.818359 -0.002415  0.013649   

            Amount  Class  
0       149.625000      0  
1         2.689453      0  
2       378.750000      0  
3       123.500000      0  
4        70.000000      0  
...            ...    ...  
284802    0.770020      0  
284803   24.796875      0  
284804   67.875000      0  
284805   10.000000      0  
284806  217.000000      0  

[284807 rows x 31 columns]

In [129]:
df['V1'].value_counts(normalize=True)

 2.062500    0.002201
 2.052734    0.001991
 2.042969    0.001864
 2.039062    0.001850
 2.054688    0.001784
               ...   
-0.015327    0.000004
 0.000822    0.000004
 0.017380    0.000004
 0.005089    0.000004
 0.018066    0.000004
Name: V1, Length: 17239, dtype: float64

In [130]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float32
 1   V1      284807 non-null  float16
 2   V2      284807 non-null  float16
 3   V3      284807 non-null  float16
 4   V4      284807 non-null  float16
 5   V5      284807 non-null  float16
 6   V6      284807 non-null  float16
 7   V7      284807 non-null  float16
 8   V8      284807 non-null  float16
 9   V9      284807 non-null  float16
 10  V10     284807 non-null  float16
 11  V11     284807 non-null  float16
 12  V12     284807 non-null  float16
 13  V13     284807 non-null  float16
 14  V14     284807 non-null  float16
 15  V15     284807 non-null  float16
 16  V16     284807 non-null  float16
 17  V17     284807 non-null  float16
 18  V18     284807 non-null  float16
 19  V19     284807 non-null  float16
 20  V20     284807 non-null  float16
 21  V21     28

In [131]:
pd.options.display.max_columns = 100

In [132]:
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359375  -0.072754  2.537109  1.377930 -0.338379   
1            0.0   1.191406   0.266113  0.166504  0.448242  0.060028   
2            1.0  -1.358398  -1.339844  1.773438  0.379883 -0.503418   
3            1.0  -0.966309  -0.185181  1.792969 -0.863281 -0.010307   
4            2.0  -1.158203   0.877930  1.548828  0.403076 -0.407227   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.882812  10.070312 -9.835938 -2.066406 -5.363281   
284803  172787.0  -0.732910  -0.055084  2.035156 -0.738770  0.868164   
284804  172788.0   1.919922  -0.301270 -3.250000 -0.557617  2.630859   
284805  172788.0  -0.240479   0.530273  0.702637  0.689941 -0.377930   
284806  172792.0  -0.533203  -0.189697  0.703125 -0.506348 -0.012543   

              V6        V7        V8        V9       V10       V11       V12  \
0       0.462402  0.239624  0.098694  0.363770  0.090820 -0.551758 -0.617676   
1      -0.082336 -0.078796  0.085083 -0.255371 -0.166992  1.612305  1.065430   
2       1.800781  0.791504  0.247681 -1.514648  0.207642  0.624512  0.066101   
3       1.247070  0.237549  0.377441 -1.386719 -0.054962 -0.226440  0.178223   
4       0.095947  0.592773 -0.270508  0.817871  0.752930 -0.822754  0.538086   
...          ...       ...       ...       ...       ...       ...       ...   
284802 -2.607422 -4.917969  7.304688  1.914062  4.355469 -1.592773  2.712891   
284803  1.058594  0.024323  0.294922  0.584961 -0.976074 -0.150146  0.916016   
284804  3.031250 -0.296875  0.708496  0.432373 -0.484863  0.411621  0.063110   
284805  0.623535 -0.686035  0.679199  0.392090 -0.399170 -1.933594 -0.962891   
284806 -0.649414  1.577148 -0.414551  0.486084 -0.915527 -1.040039 -0.031525   

             V13       V14       V15       V16       V17       V18       V19  \
0      -0.991211 -0.311279  1.467773 -0.470459  0.208008  0.025787  0.404053   
1       0.489014 -0.143799  0.635742  0.463867 -0.114807 -0.183350 -0.145752   
2       0.717285 -0.165894  2.345703 -2.890625  1.110352 -0.121338 -2.261719   
3       0.507812 -0.287842 -0.631348 -1.059570 -0.684082  1.965820 -1.232422   
4       1.345703 -1.120117  0.175171 -0.451416 -0.237061 -0.038208  0.803711   
...          ...       ...       ...       ...       ...       ...       ...   
284802 -0.689453  4.625000 -0.924316  1.107422  1.991211  0.510742 -0.683105   
284803  1.214844 -0.675293  1.165039 -0.711914 -0.025696 -1.220703 -1.545898   
284804 -0.183716 -0.510742  1.329102  0.140747  0.313477  0.395752 -0.577148   
284805 -1.041992  0.449707  1.962891 -0.608398  0.509766  1.114258  2.898438   
284806 -0.188110 -0.084290  0.041321 -0.302734 -0.660156  0.167480 -0.256104   

             V20       V21       V22       V23       V24       V25       V26  \
0       0.251465 -0.018311  0.277832 -0.110474  0.066956  0.128540 -0.189087   
1      -0.069092 -0.225830 -0.638672  0.101318 -0.339844  0.167114  0.125854   
2       0.524902  0.248047  0.771484  0.909180 -0.689453 -0.327637 -0.139038   
3      -0.208008 -0.108276  0.005272 -0.190308 -1.175781  0.647461 -0.221924   
4       0.408447 -0.009430  0.798340 -0.137451  0.141235 -0.206055  0.502441   
...          ...       ...       ...       ...       ...       ...       ...   
284802  1.475586  0.213501  0.111877  1.014648 -0.509277  1.436523  0.250000   
284803  0.059601  0.214233  0.924316  0.012466 -1.016602 -0.606445 -0.395264   
284804  0.001396  0.232056  0.578125 -0.037506  0.640137  0.265625 -0.087341   
284805  0.127441  0.265137  0.799805 -0.163330  0.123230 -0.569336  0.546875   
284806  0.383057  0.260986  0.643066  0.376709  0.008797 -0.473633 -0.818359   

             V27       V28      Amount  Class  
0       0.133545 -0.021057  149.625000      0  
1      -0.008980  0.014725    2.689453      0  
2      -0.055359 -0.059753  378.750000      0  
3       0.062744  0.061462  123.500000      0  
4       0.219482  0.2152

In [134]:
df.head(11)

Time        V1        V2        V3        V4        V5        V6  \
0    0.0 -1.359375 -0.072754  2.537109  1.377930 -0.338379  0.462402   
1    0.0  1.191406  0.266113  0.166504  0.448242  0.060028 -0.082336   
2    1.0 -1.358398 -1.339844  1.773438  0.379883 -0.503418  1.800781   
3    1.0 -0.966309 -0.185181  1.792969 -0.863281 -0.010307  1.247070   
4    2.0 -1.158203  0.877930  1.548828  0.403076 -0.407227  0.095947   
5    2.0 -0.426025  0.960449  1.140625 -0.168213  0.420898 -0.029724   
6    4.0  1.229492  0.140991  0.045380  1.202148  0.191895  0.272705   
7    7.0 -0.644043  1.417969  1.074219 -0.492188  0.948730  0.428223   
8    7.0 -0.894043  0.286133 -0.113220 -0.271484  2.669922  3.722656   
9    9.0 -0.338379  1.119141  1.043945 -0.222168  0.499268 -0.246704   
10  10.0  1.449219 -1.176758  0.914062 -1.375977 -1.971680 -0.629395   

          V7        V8        V9       V10       V11       V12       V13  \
0   0.239624  0.098694  0.363770  0.090820 -0.551758 -0.617676 -0.991211   
1  -0.078796  0.085083 -0.255371 -0.166992  1.612305  1.065430  0.489014   
2   0.791504  0.247681 -1.514648  0.207642  0.624512  0.066101  0.717285   
3   0.237549  0.377441 -1.386719 -0.054962 -0.226440  0.178223  0.507812   
4   0.592773 -0.270508  0.817871  0.752930 -0.822754  0.538086  1.345703   
5   0.476318  0.260254 -0.568848 -0.371338  1.340820  0.359863 -0.358154   
6  -0.005157  0.081238  0.464844 -0.099243 -1.416992 -0.153809 -0.750977   
7   1.121094 -3.808594  0.615234  1.249023 -0.619629  0.291504  1.757812   
8   0.370117  0.851074 -0.392090 -0.410400 -0.705078 -0.110474 -0.286133   
9   0.651367  0.069519 -0.736816 -0.366943  1.017578  0.836426  1.006836   
10 -1.422852  0.048462 -1.720703  1.626953  1.199219 -0.671387 -0.514160   

         V14       V15       V16       V17       V18       V19       V20  \
0  -0.311279  1.467773 -0.470459  0.208008  0.025787  0.404053  0.251465   
1  -0.143799  0.635742  0.463867 -0.114807 -0.183350 -0.145752 -0.069092   
2  -0.165894  2.345703 -2.890625  1.110352 -0.121338 -2.261719  0.524902   
3  -0.287842 -0.631348 -1.059570 -0.684082  1.965820 -1.232422 -0.208008   
4  -1.120117  0.175171 -0.451416 -0.237061 -0.038208  0.803711  0.408447   
5  -0.137085  0.517578  0.401611 -0.058136  0.068665 -0.033203  0.084961   
6   0.167358  0.050140 -0.443604  0.002821 -0.611816 -0.045563 -0.219604   
7  -1.324219  0.686035 -0.076111 -1.221680 -0.358154  0.324463 -0.156738   
8   0.074341 -0.328857 -0.210083 -0.499756  0.118774  0.570312  0.052734   
9  -0.443604  0.150269  0.739258 -0.541016  0.476562  0.451660  0.203735   
10 -0.095032  0.230957  0.031982  0.253418  0.854492 -0.221313 -0.387207   

         V21       V22       V23       V24       V25       V26       V27  \
0  -0.018311  0.277832 -0.110474  0.066956  0.128540 -0.189087  0.133545   
1  -0.225830 -0.638672  0.101318 -0.339844  0.167114  0.125854 -0.008980   
2   0.248047  0.771484  0.909180 -0.689453 -0.327637 -0.139038 -0.055359   
3  -0.108276  0.005272 -0.190308 -1.175781  0.647461 -0.221924  0.062744   
4  -0.009430  0.798340 -0.137451  0.141235 -0.206055  0.502441  0.219482   
5  -0.208252 -0.560059 -0.026398 -0.371338 -0.232788  0.105896  0.253906   
6  -0.167725 -0.270752 -0.154053 -0.780273  0.750000 -0.257324  0.034515   
7   1.943359 -1.015625  0.057495 -0.649902 -0.415283 -0.051636 -1.207031   
8  -0.073425 -0.268066 -0.204224  1.011719  0.373291 -0.384277  0.011749   
9  -0.246948 -0.633789 -0.120789 -0.385010 -0.069763  0.094177  0.246216   
10 -0.009300  0.313965  0.027740  0.500488  0.251465 -0.129517  0.042847   

         V28      Amount  Class  
0  -0.021057  149.625000      0  
1   0.014725    2.689453      0  
2  -0.059753  378.750000      0  
3   0.061462  123.500000      0  
4   0.215210   70.000000      0  
5   0.081055    3.669922      0  
6   0.005169    4.988281      0  
7  -1.084961   40.812500      0  
8   0.142456   93.187500      0  
9   0.083069    3.679688      0  
10  0.016251    7.800781  

In [136]:
X = df.drop(['Class'], axis = 1)
X

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359375  -0.072754  2.537109  1.377930 -0.338379   
1            0.0   1.191406   0.266113  0.166504  0.448242  0.060028   
2            1.0  -1.358398  -1.339844  1.773438  0.379883 -0.503418   
3            1.0  -0.966309  -0.185181  1.792969 -0.863281 -0.010307   
4            2.0  -1.158203   0.877930  1.548828  0.403076 -0.407227   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.882812  10.070312 -9.835938 -2.066406 -5.363281   
284803  172787.0  -0.732910  -0.055084  2.035156 -0.738770  0.868164   
284804  172788.0   1.919922  -0.301270 -3.250000 -0.557617  2.630859   
284805  172788.0  -0.240479   0.530273  0.702637  0.689941 -0.377930   
284806  172792.0  -0.533203  -0.189697  0.703125 -0.506348 -0.012543   

              V6        V7        V8        V9       V10       V11       V12  \
0       0.462402  0.239624  0.098694  0.363770  0.090820 -0.551758 -0.617676   
1      -0.082336 -0.078796  0.085083 -0.255371 -0.166992  1.612305  1.065430   
2       1.800781  0.791504  0.247681 -1.514648  0.207642  0.624512  0.066101   
3       1.247070  0.237549  0.377441 -1.386719 -0.054962 -0.226440  0.178223   
4       0.095947  0.592773 -0.270508  0.817871  0.752930 -0.822754  0.538086   
...          ...       ...       ...       ...       ...       ...       ...   
284802 -2.607422 -4.917969  7.304688  1.914062  4.355469 -1.592773  2.712891   
284803  1.058594  0.024323  0.294922  0.584961 -0.976074 -0.150146  0.916016   
284804  3.031250 -0.296875  0.708496  0.432373 -0.484863  0.411621  0.063110   
284805  0.623535 -0.686035  0.679199  0.392090 -0.399170 -1.933594 -0.962891   
284806 -0.649414  1.577148 -0.414551  0.486084 -0.915527 -1.040039 -0.031525   

             V13       V14       V15       V16       V17       V18       V19  \
0      -0.991211 -0.311279  1.467773 -0.470459  0.208008  0.025787  0.404053   
1       0.489014 -0.143799  0.635742  0.463867 -0.114807 -0.183350 -0.145752   
2       0.717285 -0.165894  2.345703 -2.890625  1.110352 -0.121338 -2.261719   
3       0.507812 -0.287842 -0.631348 -1.059570 -0.684082  1.965820 -1.232422   
4       1.345703 -1.120117  0.175171 -0.451416 -0.237061 -0.038208  0.803711   
...          ...       ...       ...       ...       ...       ...       ...   
284802 -0.689453  4.625000 -0.924316  1.107422  1.991211  0.510742 -0.683105   
284803  1.214844 -0.675293  1.165039 -0.711914 -0.025696 -1.220703 -1.545898   
284804 -0.183716 -0.510742  1.329102  0.140747  0.313477  0.395752 -0.577148   
284805 -1.041992  0.449707  1.962891 -0.608398  0.509766  1.114258  2.898438   
284806 -0.188110 -0.084290  0.041321 -0.302734 -0.660156  0.167480 -0.256104   

             V20       V21       V22       V23       V24       V25       V26  \
0       0.251465 -0.018311  0.277832 -0.110474  0.066956  0.128540 -0.189087   
1      -0.069092 -0.225830 -0.638672  0.101318 -0.339844  0.167114  0.125854   
2       0.524902  0.248047  0.771484  0.909180 -0.689453 -0.327637 -0.139038   
3      -0.208008 -0.108276  0.005272 -0.190308 -1.175781  0.647461 -0.221924   
4       0.408447 -0.009430  0.798340 -0.137451  0.141235 -0.206055  0.502441   
...          ...       ...       ...       ...       ...       ...       ...   
284802  1.475586  0.213501  0.111877  1.014648 -0.509277  1.436523  0.250000   
284803  0.059601  0.214233  0.924316  0.012466 -1.016602 -0.606445 -0.395264   
284804  0.001396  0.232056  0.578125 -0.037506  0.640137  0.265625 -0.087341   
284805  0.127441  0.265137  0.799805 -0.163330  0.123230 -0.569336  0.546875   
284806  0.383057  0.260986  0.643066  0.376709  0.008797 -0.473633 -0.818359   

             V27       V28      Amount  
0       0.133545 -0.021057  149.625000  
1      -0.008980  0.014725    2.689453  
2      -0.055359 -0.059753  378.750000  
3       0.062744  0.061462  123.500000  
4       0.219482  0.215210   70.000000  
...          ...  

In [138]:
y = df['Class']

In [140]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=100, stratify=y)

In [144]:
X_train.shape, y_train.shape, X_test.shape, y_test.shape


((199364, 30), (199364,), (85443, 30), (85443,))

In [145]:
parameters = [{'n_estimators': [10, 15],
'max_features': np.arange(3, 5),
'max_depth': np.arange(4, 7)}]


In [146]:
 clf = GridSearchCV(estimator=RandomForestClassifier(random_state=100), 
                   param_grid=parameters, 
                   scoring='roc_auc', 
                   cv=3)

In [147]:
clf.fit(X_train, y_train)

GridSearchCV(cv=3, estimator=RandomForestClassifier(random_state=100),
             param_grid=[{'max_depth': array([4, 5, 6]),
                          'max_features': array([3, 4]),
                          'n_estimators': [10, 15]}],
             scoring='roc_auc')

In [148]:
clf.best_params_

{'max_depth': 5, 'max_features': 4, 'n_estimators': 15}

In [151]:
y_proba = clf.predict_proba(X_test)

In [156]:
y_pred_proba = y_proba[:, 1]

In [157]:
y_pred_proba

array([0.00065102, 0.00037951, 0.00033822, ..., 0.00033822, 0.00065098,
       0.00168331])

In [158]:
from sklearn.metrics import roc_auc_score

In [159]:
roc_auc_score(y_test, y_pred_proba)

0.9494955504188167